drivers/serial/uart_mspm0.c

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright (c) 2024 Texas Instruments
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ti_mspm0_uart
+
+/* Zephyr includes */
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/mspm0_clock_control.h>
+#include <zephyr/irq.h>
+#include <soc.h>
+
+/* Driverlib includes */
+#include <ti/driverlib/dl_uart_main.h>
+
+struct uart_mspm0_config {
+	UART_Regs *regs;
+	uint32_t clock_frequency;
+	uint32_t current_speed;
+	const struct mspm0_clockSys *clock_subsys;
+	const struct pinctrl_dev_config *pinctrl;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_config_func_t irq_config_func;
+#endif
+};
+
+struct uart_mspm0_data {
+	/* UART clock structure */
+	DL_UART_Main_ClockConfig UART_ClockConfig;
+	/* UART config structure */
+	DL_UART_Main_Config UART_Config;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uint32_t interruptState;          /* Masked Interrupt Status when called by irq_update */
+	uart_irq_callback_user_data_t cb; /* Callback function pointer */
+	void *cb_data;                    /* Callback function arg */
+#endif                                    /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static void uart_mspm0_isr(const struct device *dev);
+#define MSPM0_INTERRUPT_CALLBACK_FN(index) .cb = NULL,
+#else
+#define MSPM0_INTERRUPT_CALLBACK_FN(index)
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+static int uart_mspm0_init(const struct device *dev)
+{
+	const struct uart_mspm0_config *config = dev->config;
+	struct uart_mspm0_data *data = dev->data;
+	const struct device *const clk_dev = DEVICE_DT_GET(DT_NODELABEL(clkmux));
+	uint32_t clock_rate;
+	int ret;
+
+	/* Reset power */
+	DL_UART_Main_reset(config->regs);
+	DL_UART_Main_enablePower(config->regs);
+	delay_cycles(POWER_STARTUP_DELAY);
```

Comment:
It works fine without this delay in my end. Why do we need this delay in CPU cycles?

Reply:
It's part of the settling time of peripherals. It may be possible to shrink them, but this is the way to make sure that all peripherals have settling time. It does add to the time significantly, which is why in baremetal there was a pre-init for all power-on with one delay, and then we didn't have to do the delay each time

Reply:
Please share the TRM or datasheet or document which is mandating it

Reply:
Understood from the other comment. But let's keep it 4